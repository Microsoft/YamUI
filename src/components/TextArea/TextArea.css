/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import "src/css/variables/colors.css";
@import "src/css/variables/fonts.css";

// work is in progress to set styling/theme options for all Fabric label instances
.y-text-field .ms-Label {
  @extend text__mediumSub;
  font-weight: $fontWeight__bold;
  color: $textColor__secondary;
  padding-bottom: .4rem;
}

.y-text-field .ms-TextField-fieldGroup {
  line-height: 3.3rem;
  height: 3.6rem;
}

.y-text-field .ms-TextField-field {
  @extend text__mediumSub;
  color: $textColor;
  padding: 0 .8rem;
}

.y-text-field.ms-TextField--multiline .ms-TextField-field {
  padding-top: 10px;
  padding-bottom: 7px;
}

.y-text-field.is-disabled .ms-TextField-field {
  color: $textColor__disabled;
}

.y-text-field .ms-TextField-description {
  @extend text__small;
  color: $textColor__metadata;
}

.y-text-field .ms-TextField-errorMessage {
  @extend text__small;
  color: $textColor__error;
}

/* Hide the asterisk that comes with required=true */
.y-text-field .ms-Label::after,
.y-text-field .ms-TextField-fieldGroup::after {
  display: none;
}
