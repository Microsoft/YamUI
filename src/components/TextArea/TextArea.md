### Examples

Basic
```js { "props": { "data-example": "basic" } }
<TextArea
  value="This multiline has fixed height so scrollbars may appear. This multiline has fixed height so scrollbars may appear. This multiline has fixed height so scrollbars may appear. This multiline has fixed height so scrollbars may appear. This multiline has fixed height so scrollbars may appear. This multiline has fixed height so scrollbars may appear."
  rows={3}
/>
```

With a description
```js { "props": { "data-example": "withDescription" } }
<TextArea
  value="This multiline has fixed height so scrollbars may appear."
  rows={3}
  description="Small hint goes here"
/>
```

Auto-adjust Height
```js { "props": { "data-example": "autoAdjustHeight" } }
<TextArea
  value="This multiline auto grows. This multiline auto grows.  This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows. This multiline auto grows."
  autoAdjustHeight={true}
  rows={1}
/>
```

Disabled
```js { "props": { "data-example": "disabled" } }
<TextArea
  value="This is disabled."
  disabled={true}
/>
```

Placeholder
```js { "props": { "data-example": "placeHolder" } }
<TextArea
  placeHolder="This is a placeholder."
/>
```

Error
```js { "props": { "data-example": "error" } }
<TextArea
  value="This value has an error."
  errorMessage="Error Explanation"
  rows={3}
/>
```