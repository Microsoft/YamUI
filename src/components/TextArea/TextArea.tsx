/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { TextField as FabricTextField } from 'office-ui-fabric-react/lib/TextField';
import { Async } from 'office-ui-fabric-react/lib/Utilities';
import './TextArea.css';

export interface TextAreaProps extends BaseComponentProps {
  /**
   * Current value of the TextArea.
  */
  value?: string;

  /**
   * Placeholder text. Displayed until a value is entered.
  **/
  placeHolder?: string;

  /**
   * Error message string.
   */
  errorMessage?: string;

  /**
   * Disabled state of the TextArea.
  */
  disabled?: boolean;

  /**
   * Label for the TextArea.
  */
  label?: string;

  /**
   * The TextArea input description
   */
  description?: string;

  /**
   * Use to indicate that a value must be provided to allow the containing form to be submitted.
  */
  required?: boolean;

  /**
   * Height of the field, in number of rows.
  **/
  rows?: number;

  /**
   * Whether or not to auto adjust textField height.
  **/
  autoAdjustHeight?: boolean;

  /**
   * Callback for the onChanged event.
   */
  onChange?: (newValue: any) => void;

  /**
   * TextArea will trigger `onChange` after users stop typing for `onChangeDebounceTime`
   * milliseconds.
   */
  onChangeDebounceTime?: number;

  /**
   * Focus callback handler.
   */
  onFocus?: (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => void;

  /**
   * Blur callback handler.
   */
  onBlur?: (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => void;

  /**
   * Mouse enter callback handler.
   */
  onMouseEnter?: (event: React.MouseEvent<HTMLInputElement | HTMLTextAreaElement>) => void;

  /**
   * Mouse leave callback handler.
   */
  onMouseLeave?: (event: React.MouseEvent<HTMLInputElement | HTMLTextAreaElement>) => void;
}

/**
 * The TextArea component enables a user to type text into an app. It's typically used to capture
 * a single line of text, but can be configured to capture multiple lines of text. The text
 * displays on the screen in a simple, uniform format.
 */
export default class TextArea extends React.PureComponent<TextAreaProps, {}> {
  private async: Async;
  private debouncedOnChange: (newValue: string) => void;

  constructor(props: TextAreaProps) {
    super();
    if (props.onChangeDebounceTime && props.onChange) {
      this.async = new Async(this);
      this.debouncedOnChange = this.async.debounce(props.onChange, props.onChangeDebounceTime);
    }
  }

  componentWillUnmount() {
    this.async.dispose();
  }

  render() {
    return (
      <FabricTextField
        value={this.props.value}
        className={this.getClasses()}
        label={this.props.label}
        description={this.props.errorMessage ? undefined : this.props.description}
        disabled={this.props.disabled}
        required={this.props.required}
        errorMessage={this.props.errorMessage}
        placeholder={this.props.placeHolder}
        rows={this.props.rows}
        autoAdjustHeight={this.props.autoAdjustHeight}
        multiline={true}
        resizable={false}
        onChanged={this.debouncedOnChange || this.props.onChange}
        onFocus={this.props.onFocus}
        onBlur={this.props.onBlur}
        onMouseEnter={this.props.onMouseEnter}
        onMouseLeave={this.props.onMouseLeave}
      />
    );
  }

  private getClasses() {
    const { className } = this.props;
    const classes = ['y-text-field'];

    if (className) {
      classes.push(className);
    }

    return classes.join(' ');
  }
}
