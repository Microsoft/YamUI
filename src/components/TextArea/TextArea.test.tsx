/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */

import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import TextArea, { TextAreaProps } from '.';
import { TextField as FabricTextField, ITextFieldProps }
  from 'office-ui-fabric-react/lib/TextField';
import { Async } from 'office-ui-fabric-react/lib/Utilities';

describe('<TextArea />', () => {
  let component: ShallowWrapper<TextAreaProps, {}>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<TextArea />);
    });

    it('has its correct base class', () => {
      expect(component.hasClass('y-text-field')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with all options', () => {
    beforeEach(() => {
      component = shallow(
        <TextArea
          className="CLASS"
          value="VALUE"
          placeHolder="PLACEHOLDER"
          disabled={false}
          label="LABEL"
          description="DESCRIPTION"
          rows={2}
          autoAdjustHeight={true}
          required={true}
          onChange={jest.fn()}
          onChangeDebounceTime={500}
        />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('unmount', () => {
      let disposeSpy: jest.SpyInstance<() => void>;

      beforeEach(() => {
        disposeSpy = jest.spyOn(Async.prototype, 'dispose');
        component.unmount();
      });

      it('calls async dispose', () => {
        expect(disposeSpy).toHaveBeenCalled();
      });
    });
  });

  describe('with error message and description', () => {
    beforeEach(() => {
      component = shallow(
        <TextArea
          description="DESCRIPTION"
          errorMessage="ERROR"
        />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with event callbacks', () => {
    let onMouseEnter: jest.Mock<{}>;
    let onMouseLeave: jest.Mock<{}>;
    let onFocus: jest.Mock<{}>;
    let onBlur: jest.Mock<{}>;

    beforeEach(() => {
      onMouseEnter = jest.fn();
      onMouseLeave = jest.fn();
      onFocus = jest.fn();
      onBlur = jest.fn();
      component = shallow(
        <TextArea
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          onFocus={onFocus}
          onBlur={onBlur}
        />,
      );
    });

    describe('when hovered', () => {
      beforeEach(() => {
        component.simulate('mouseEnter');
      });

      it('triggers its onMouseEnter callback', () => {
        expect(onMouseEnter).toHaveBeenCalled();
      });
    });

    describe('when hovered out', () => {
      beforeEach(() => {
        component.simulate('mouseLeave');
      });

      it('triggers its onMouseLeave callback', () => {
        expect(onMouseLeave).toHaveBeenCalled();
      });
    });

    describe('when focused', () => {
      beforeEach(() => {
        component.simulate('focus');
      });

      it('triggers its onFocus callback', () => {
        expect(onFocus).toHaveBeenCalled();
      });
    });

    describe('when blurred', () => {
      beforeEach(() => {
        component.simulate('blur');
      });

      it('triggers its onBlur callback', () => {
        expect(onBlur).toHaveBeenCalled();
      });
    });
  });

  describe('with onChange handler', () => {
    let callback: jest.Mock<{}>;

    beforeEach(() => {
      callback = jest.fn();
      component = shallow(<TextArea onChange={callback} />);
    });

    describe('when a change in made', () => {
      let fabricTextArea: ShallowWrapper<ITextFieldProps, {}>;

      beforeEach(() => {
        fabricTextArea = component.find(FabricTextField);
        fabricTextArea.prop<Function>('onChanged')('a');
      });

      it('triggers the onChange callback', () => {
        expect(callback).toHaveBeenCalledWith('a');
      });
    });
  });
});
