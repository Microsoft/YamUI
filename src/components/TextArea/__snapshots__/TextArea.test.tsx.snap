// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> with all options matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={true}
  borderless={false}
  className="y-text-field CLASS"
  deferredValidationTime={200}
  description="DESCRIPTION"
  disabled={false}
  errorMessage=""
  label="LABEL"
  multiline={true}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  placeholder="PLACEHOLDER"
  required={true}
  resizable={false}
  rows={2}
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
  value="VALUE"
/>
`;

exports[`<TextArea /> with default options matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-text-field"
  deferredValidationTime={200}
  errorMessage=""
  multiline={true}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  resizable={false}
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
/>
`;

exports[`<TextArea /> with error message and description matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-text-field"
  deferredValidationTime={200}
  errorMessage="ERROR"
  multiline={true}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  resizable={false}
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
/>
`;
